presto-raptor/src/test/java/com/facebook/presto/raptor/storage/organization/TestShardOrganizationManager.java

```diff
@@ -0,0 +1,197 @@
+import static com.google.common.collect.Iterables.getOnlyElement;
+import static io.airlift.units.Duration.valueOf;
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+import static java.util.stream.Collectors.toSet;
+import static org.testng.Assert.assertEquals;
+
+@Test(singleThreaded = true)
+public class TestShardOrganizationManager
+{
+    private IDBI dbi;
+    private Handle dummyHandle;
+    private MetadataDao metadataDao;
+    private ShardOrganizerDao organizerDao;
+
+    private static final Table tableInfo = new Table(1L, OptionalLong.empty(), OptionalInt.empty(), OptionalLong.empty());
+    private static final Table temporalTableInfo = new Table(1L, OptionalLong.empty(), OptionalInt.empty(), OptionalLong.of(1));
+
+    private static final List<Type> types = ImmutableList.of(BIGINT, VARCHAR, DATE, TIMESTAMP);
+
+    @BeforeMethod
+    public void setup()
+    {
+        dbi = new DBI("jdbc:h2:mem:test" + System.nanoTime());
+        dummyHandle = dbi.open();
+        metadataDao = dbi.onDemand(MetadataDao.class);
+        organizerDao = dbi.onDemand(ShardOrganizerDao.class);
+
+        createTablesWithRetry(dbi);
+    }
+
+    @AfterMethod
+    public void teardown()
+    {
+        dummyHandle.close();
+    }
+
+    @Test
+    public void testOrganizationEligibleTables()
+            throws Exception
+    {
+        long table1 = metadataDao.insertTable("schema", "table1", false, true, null);
+        metadataDao.insertColumn(table1, 1, "foo", 1, "bigint", 1, null);
+
+        metadataDao.insertTable("schema", "table2", false, true, null);
+        metadataDao.insertTable("schema", "table3", false, false, null);
+        assertEquals(metadataDao.getOrganizationEligibleTables(), ImmutableSet.of(table1));
+    }
+
+    @Test(timeOut = 1_000)
+    public void testTableDiscovery()
+            throws Exception
+    {
+        long table1 = metadataDao.insertTable("schema", "table1", false, true, null);
+        metadataDao.insertColumn(table1, 1, "foo", 1, "bigint", 1, null);
+
+        long table2 = metadataDao.insertTable("schema", "table2", false, true, null);
+        metadataDao.insertColumn(table2, 1, "foo", 1, "bigint", 1, null);
+
+        metadataDao.insertTable("schema", "table3", false, false, null);
+
+        ShardOrganizationManager organizationManager = createShardOrganizationManager("100ms");
+
+        // initializes tables
+        Set<Long> actual = organizationManager.discoverAndInitializeTablesToOrganize();
+        assertEquals(actual, ImmutableSet.of(table1, table2));
+
+        // update the start times and test that the tables are discovered after interval seconds
+        organizerDao.updateLastStartTime("node1", table1);
+        organizerDao.updateLastStartTime("node1", table2);
+
+        // wait for some time (interval time) for the tables to be eligible for organization
+        while (organizationManager.discoverAndInitializeTablesToOrganize().isEmpty()) {
```

Comment:
Using `@Test(timeOut = ...)` causes TestNG to run the test in a separate thread. However, the thread running the test could be stuck in the loop forever. Instead, make this loop for up to a second using `System.nanoTime()` so it's guaranteed to exit.